lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -332,6 +332,94 @@ LogicalResult ChannelBufferLowering::matchAndRewrite(
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ESIPureModuleOp pureMod, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final;
+};
+} // anonymous namespace
+
+LogicalResult
+PureModuleLowering::matchAndRewrite(ESIPureModuleOp pureMod, OpAdaptor adaptor,
+                                    ConversionPatternRewriter &rewriter) const {
+  auto loc = pureMod.getLoc();
+  Block *body = &pureMod.getBody().front();
+
+  // Track existing names (so we can de-dup) and get op result when we want to
+  // replace it with the block args.
+  DenseMap<StringAttr, ESIPureModuleInputOp> inputPortNames;
+  // Build the port list for `hw.module` construction.
+  SmallVector<hw::PortInfo> ports;
+  // List the input and output ops.
+  SmallVector<ESIPureModuleInputOp> inputs;
+  SmallVector<ESIPureModuleOutputOp> outputs;
+
+  for (Operation &op : llvm::make_early_inc_range(body->getOperations())) {
+    if (auto port = dyn_cast<ESIPureModuleInputOp>(op)) {
+      // If we already have an input port of the same name, replace the result
+      // value with the previous one. Checking that the types match is done in
+      // the pure module verifier.
+      auto existingPort = inputPortNames.find(port.getNameAttr());
+      if (existingPort != inputPortNames.end()) {
+        rewriter.replaceAllUsesWith(port.getResult(),
+                                    existingPort->getSecond().getResult());
+        rewriter.eraseOp(port);
+        continue;
+      }
+      // Normal port construction.
+      ports.push_back(hw::PortInfo{port.getNameAttr(),
+                                   hw::PortDirection::INPUT,
+                                   port.getResult().getType(),
+                                   inputs.size(),
+                                   {},
+                                   port.getLoc()});
+      inputs.push_back(port);
+    } else if (auto port = dyn_cast<ESIPureModuleOutputOp>(op)) {
+      ports.push_back(hw::PortInfo{port.getNameAttr(),
+                                   hw::PortDirection::OUTPUT,
+                                   port.getValue().getType(),
+                                   outputs.size(),
+                                   {},
+                                   port.getLoc()});
+      outputs.push_back(port);
+    }
+  }
+
+  // Create the replacement `hw.module`.
+  auto hwMod =
+      rewriter.create<hw::HWModuleOp>(loc, pureMod.getNameAttr(), ports);
+  rewriter.eraseBlock(hwMod.getBodyBlock());
+  rewriter.inlineRegionBefore(*body->getParent(), hwMod.getBodyRegion(),
+                              hwMod.getBodyRegion().end());
+  body = hwMod.getBodyBlock();
+
+  // Re-wire the inputs and erase them.
+  for (auto input : inputs) {
+    rewriter.replaceAllUsesWith(
+        input.getResult(),
+        body->addArgument(input.getResult().getType(), input.getLoc()));
```

Comment:
Is there a `rewriter` API for adding a block argument? If not, I think pedantically this should be wrapped in `rewriter.updateRootInPlace(body, [](){ body->addArgument(...); })`.

Reply:
Yep. It's this sort of thing that I wanted a review for. Thanks!